### PR TITLE
Added Image Geometry class

### PR DESCRIPTION
**Pull request by andrewstanfordjason**

### CODE REVIEW

File: lib_nn/api/geom/ImageGeometry.hpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+  T &Element(T *img_base, const int row, const int col,
+             const int channel) const;
+
+  /**
+   * Get the value of the specified element from the provided image.
+   * If the coordinates refer to a location outside the image geometry pad_value
+   * is returned.
+   */
+  template <typename T>
+  T Get(const T *img_base, const ImageVect coords, const T pad_value = 0) const;
```
Comment (ghost) [inline lib_nn/api/geom/ImageGeometry.hpp]: Did you mean `coords` to be `&coords` instead?

File: lib_nn/api/geom/ImageGeometry.hpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+   * The total number of bytes of the image
+   */
+  int inline const ImageBytes() const { return ElementCount() * channel_depth; }
+
+  /**
+   * Get the flattened index of the specified image element.
+   *
+   * The "flattened" index of an element is the index of the element when the
+   * image is stored in a 1 dimensional array. This is ideal, for example, when
+   * the image image is backed by a `std::vector` object.
```
Comment (ghost) [inline lib_nn/api/geom/ImageGeometry.hpp]:
```suggestion
   * the image is backed by a `std::vector` object.
```

File: lib_nn/api/geom/ImageGeometry.hpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+/////////////////////////
+inline std::ostream &operator<<(std::ostream &stream,
+                                const ImageGeometry &image) {
+  return stream << image.height << "," << image.width << "," << image.depth
+                << "," << image.channel_depth;
+}
+
+}  // namespace nn
```
Comment (ghost) [inline lib_nn/api/geom/ImageGeometry.hpp]: Could you please the files to have an extra line at the end?

File: lib_nn/api/geom/ImageGeometry.hpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+   */
+  bool operator==(ImageGeometry other) const;
+};
+
+/////////////////////////
+template <typename T>
+T &ImageGeometry::Element(T *img_base, const int row, const int col,
+                          const int channel) const {
+  assert(IsWithinImage(row, col, channel));
+  int index = row * this->width * this->depth + col * this->depth + channel;
```
Comment (ghost) [inline lib_nn/api/geom/ImageGeometry.hpp:201]: I'm not sure how the calculation is meant to be done. Is `this->height` missing here?

File: lib_nn/api/geom/ImageGeometry.hpp (hunk 3 above)
Comment (andrewstanfordjason) [reply lib_nn/api/geom/ImageGeometry.hpp]: yeah, no, problem. Apparently clang-format doesn't do this for me!

File: lib_nn/src/cpp/filt2d/geom/ImageGeometry.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+using namespace nn;
+
+int ImageGeometry::Index(const int row, const int col,
+                         const int channel) const {
+  if (row < 0 || row >= this->height) return -1;
+  if (col < 0 || col >= this->width) return -1;
+  if (channel < 0 || channel >= this->depth) return -1;
+
+  return this->depth * (this->width * row + col) + channel;
```
Comment (ghost) [inline lib_nn/src/cpp/filt2d/geom/ImageGeometry.cpp:11]: Similarly, is `this->height` missing here?

File: test/shared/include/Rand.hpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
+#pragma once
```
Comment (ghost) [inline test/shared/include/Rand.hpp:1]: Are we including the Rand files from somewhere else?

File: lib_nn/api/geom/ImageGeometry.hpp (hunk 4 above)
Comment (andrewstanfordjason) [reply lib_nn/api/geom/ImageGeometry.hpp:201]: nothing is missing here as the leading dimension is not required for dereferencing

File: lib_nn/src/cpp/filt2d/geom/ImageGeometry.cpp (hunk 5 above)
Comment (andrewstanfordjason) [reply lib_nn/src/cpp/filt2d/geom/ImageGeometry.cpp:11]: as above

File: test/shared/include/Rand.hpp (hunk 6 above)
Comment (andrewstanfordjason) [reply test/shared/include/Rand.hpp:1]: no

**Pull request merged**